### PR TITLE
Grant availability metrics script

### PR DESCRIPTION
**Pull request by KPrasch**

### CODE REVIEW

File: tests/metrics/grant_availability.py [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+    message = f'{METRICS_ADDRESS_ENVVAR}, {METRICS_SIGNER_ENVVAR} and {METRICS_PASSWORD_ENVVAR}' \
+              f' are required to run grant availability metrics.'
+    raise RuntimeError(message)
+
+# Alice Configuration
+DOMAIN: str = 'ibex'
+DEFAULT_SEEDNODE_URIS: List[str] = ['https://ibex.nucypher.network:9151', ]
+ETHEREUM_PROVIDER_URI: str = 'wss://rinkeby.infura.io/ws/v3/8fb725f72a3342f6a8cf8fc7ab674123'
+INSECURE_PASSWORD: str = "METRICS_INSECURE_DEVELOPMENT_PASSWORD"
+TEMP_ALICE_DIR: str = os.path.join('/', 'tmp', 'grant-metrics')
```
Comment (vzotova) [inline tests/metrics/grant_availability.py:54]: maybe use `mktmpdir` and `Path`?

File: tests/metrics/grant_availability.py [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+    decrypting_secret = UmbralPrivateKey.gen_key()
+    decrypting_key = decrypting_secret.pubkey
+    bob = Bob.from_public_keys(verifying_key=bob_verifying_key,
+                               encrypting_key=decrypting_key,
+                               federated_only=False)
+    return bob
+
+
+def metric_grant(alice, handpicked_ursulas: Optional[Set[Ursula]] = None) -> Policy:
+    """Perform a granting operation for metreics collection."""
```
Comment (arjunhassard) [inline tests/metrics/grant_availability.py:85]:
```suggestion
    """Perform a granting operation for metrics collection."""
```

File: tests/metrics/grant_availability.py [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+
+    alice_config.initialize(password=INSECURE_PASSWORD)
+    alice_config.keyring.unlock(password=INSECURE_PASSWORD)
+    alice = alice_config.produce(client_password=SIGNER_PASSWORD)
+    alice.start_learning_loop(now=True)
+    return alice
+
+
+def setup():
+    """Prepares the filesystem and logger for grsnt metrics collection"""
```
Comment (arjunhassard) [inline tests/metrics/grant_availability.py:159]:
```suggestion
    """Prepares the filesystem and logger for grant metrics collection"""
```

File: tests/metrics/grant_availability.py [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+def aggregate_nodes() -> Tuple[Set[Ursula], Set[Ursula]]:
+    """generates ursulas from URIs used in grant metrics collection"""
+
+    seednodes = set()
+    if DEFAULT_SEEDNODE_URIS:
+        for uri in DEFAULT_SEEDNODE_URIS:
+            ursula = Ursula.from_seed_and_stake_info(seed_uri=uri, federated_only=False)
+            seednodes.add(ursula)
+
+    handpicked_ursulas = set()
```
Comment (arjunhassard) [inline tests/metrics/grant_availability.py:174]: This helps developers/Alices discover the availability of a subset of Ursulas that they specify in a URI list?

**Review by arjunhassard — APPROVED** [review]
👌

File: tests/metrics/grant_availability.py [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+    SIGNER_URI: str = os.environ[METRICS_SIGNER_ENVVAR]
+except KeyError:
+    message = f'{METRICS_ADDRESS_ENVVAR}, {METRICS_SIGNER_ENVVAR} and {METRICS_PASSWORD_ENVVAR}' \
+              f' are required to run grant availability metrics.'
+    raise RuntimeError(message)
+
+# Alice Configuration
+DOMAIN: str = 'ibex'
+DEFAULT_SEEDNODE_URIS: List[str] = ['https://ibex.nucypher.network:9151', ]
+ETHEREUM_PROVIDER_URI: str = 'wss://rinkeby.infura.io/ws/v3/8fb725f72a3342f6a8cf8fc7ab674123'
```
[the remark this answers is not in the document]
Comment (KPrasch) [reply tests/metrics/grant_availability.py:52]: Haha - yes - let's move this to an envvar. :sweat_smile:

File: tests/metrics/grant_availability.py (hunk 4 above)
Comment (KPrasch) [reply tests/metrics/grant_availability.py:174]: @arjunhassard - Yes, exactly.

**Review by cygnusv — APPROVED** [review]
Let's move this forward :)

File: tests/metrics/grant_availability.py [hunk 6]
```diff
@@ -0,0 +1,187 @@
...
+            success += 1
+            policies[policy.public_key.hex()] = policy  # track
+            print(f"PEK:{policy.public_key.hex()}")
+
+        # timeit
+        end = maya.now()
+        delta = end - start
+        print(f"Completed in {(delta).total_seconds()} seconds.")
+
+        if i == iterations and not run_forever:
```
Comment (derekpierre) [inline tests/metrics/grant_availability.py:122]: I think we are running once more than expected and the check should be `(i+1) == iterations`. We increment `i` after we run i.e. i = 0 is really iteration 1

**Review by derekpierre — CHANGES_REQUESTED** [review]

**Comment by KPrasch** [comment]
Error captured in a resent result (blocked by #1861 and friends) : https://dpaste.com/AZ8DYD36G

File: tests/metrics/grant_availability.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (derekpierre) [reply tests/metrics/grant_availability.py:52]: Still planning on doing an envvar here?

File: tests/metrics/grant_availability.py (hunk 1 above)
Comment (derekpierre) [reply tests/metrics/grant_availability.py:54]: I like @vzotova's suggestion here.

File: tests/metrics/grant_availability.py (hunk 6 above)
Comment (derekpierre) [reply tests/metrics/grant_availability.py:122]: Are we running on extra time here?

Lets say you want 1 iteration.

Loop 1: i = 0, perform metric grant, 0 != 1 so continue the loop, increment i to 1
Loop 2: i = 1, perform metric grant, 1 == 1 so stop.

**Review by derekpierre — COMMENTED** [review]
@KPrasch comments to clarify my concerns.

File: tests/metrics/grant_availability.py (hunk 6 above)
Comment (KPrasch) [reply tests/metrics/grant_availability.py:122]: I knew what you meat - thanks for explanation though :-)

I think I forgot to add these two lines to my last patch.

File: tests/metrics/grant_availability.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (KPrasch) [reply tests/metrics/grant_availability.py:52]: Heh - my computer froze when pushing, standby.

File: tests/metrics/grant_availability.py [hunk 7]
```diff
@@ -0,0 +1,187 @@
...
+        start = maya.now()
+        try:
+            policy = metric_grant(alice=alice, handpicked_ursulas=handpicked_ursulas)
+        except Exception as e:
+            fail += 1
+            print(f'GRANT FAIL\n{e}')
+        else:
+            success += 1
+            policies[policy.public_key.hex()] = policy  # track
+            print(f"PEK:{policy.public_key.hex()}")
```
[the remark this answers is not in the document]
Comment (KPrasch) [reply tests/metrics/grant_availability.py:119]: I don't see why not... the more information we can extract the better IMO.

File: tests/metrics/grant_availability.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (KPrasch) [reply tests/metrics/grant_availability.py]: :heavy_check_mark:

File: tests/metrics/grant_availability.py [hunk 8]
```diff
@@ -0,0 +1,187 @@
...
+        # timeit
+        end = maya.now()
+        delta = end - start
+        print(f"Completed in {(delta).total_seconds()} seconds.")
+
+        if i == iterations and not run_forever:
+            break  # Exit
+
+        # score
+        elif i != iterations:
```
[the remark this answers is not in the document]
Comment (KPrasch) [reply tests/metrics/grant_availability.py]: :heavy_check_mark:

File: tests/metrics/grant_availability.py (hunk 6 above)
Comment (KPrasch) [reply tests/metrics/grant_availability.py]: :heavy_check_mark:

File: tests/metrics/grant_availability.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (derekpierre) [reply tests/metrics/grant_availability.py:119]: Did you want to add the label to the text here?

**Review by derekpierre — APPROVED** [review]
🎸  - couple remaining minor comments (tmp directory creation, adding the label to the PEK output) but nothing to hold back the merge. Merge at will.

**Comment by KPrasch** [comment]
:+1: - I'm in no rush to merge this one - I'll keep adding to it until someone else merges it.

**Pull request merged**